### ISSUE TITLE
Need a way to query associated object's attributes

### ISSUE THREAD

**Issue by gregorypierce** [balderdashy/waterline#266]
If an object has an associated collection, it is not clear how (or even if its possible) to query into that associated object stuff.

Pet has an Owner
Owner has an emailAddress

How do I select all of the Pets whose owner has an `emailAddress == "foo@bar.com"`

While its easy to do 2 queries to accomplish this:
1) Find the owner who has an `emailAddress == "foo@bar.com"` and get that persons `id`
2) Find all Pets who have the `id` from query 1

It seems that there would be a better syntax for this:

```javascript
Pet.findAll({ "owner.emailAddress" : "foo@bar.com" })
```

Now whether or not this is internally optimized or just syntactic sugar doesn't matter, its much easier to read and write.

Further, if a Pet has a breed how would I select the email addresses of all of the owners whose `Pet.breed == "beagle"`

```javascript
Owner.findAll({"pet.breed" : "beagle"})
```

**Comment by Globegitter** [comment]
As I mentioned in #855 I needed this feature as well with the special case of filtering for `null` or rather `not: null`.  Which basically means I only want to find records that have a certain relation attached to them. Ideally this would work with and without actually populating that association in the original model.

**Comment by eddieajau** [comment]
Ok, so how can I help (I can start on this yesterday as it's under the "extremely desirable" label for a current project)? Is there anyone that can guide me to the first places to start looking at how to start scoping the work? In terms of API, are we talking about following what is suggested for [Waterline2](https://www.npmjs.com/package/waterline2#where-subqueries)?

Thanks in advance.

**Comment by dmarcelino** [comment]
Hey @eddieajau, your help would be most appreciated! I'm not sure how much we can reuse [waterline2](https://github.com/mikermcneil/waterline2) since a lot has changed in the last couple of months and  waterline2 hasn't been touched since December.

@devinivy, @tjwebb, @particlebanana, what are your thoughts on this?

**Comment by devinivy** [comment]
@eddieajau I'm not sure what the spec for this was in waterline2.  Does someone have a link to a proposed spec?  I like this dot notation just fine!

As for scoping work, it's fairly broad.  I think ideal first steps would be to, 

1. Write this as syntactic sugar, without any adapter-level optimizations, but perhaps keeping them in mind for the future.  This must be in place, as waterline should able to act as a shim when an adapter cannot optimize on it's own.  (Consider a key-value store such as redis trying to accomplish this task, for example.)
2. Simply _imagine_ how we can delegate optimizations to the adapter more readily.  In the long-haul, I imagine @dmarcelino's adapter "features" could become important in telling waterline what functionality it needs to shim.  See https://github.com/balderdashy/waterline-adapter-tests/pull/64.

Eventually we can support adapter-level/same-connection optimizations, but that's the hard part!  I'm sure that whomever implements the waterline shim for this feature will have some concrete insights into how to implement adapter-level optimizations.  Investigate the adapter `join` method and see how useful it might be, a la https://github.com/balderdashy/sails-mysql/blob/master/lib/adapter.js#L506.

**Comment by dmarcelino** [comment]
After reading #469 and #1030 I'm wondering if we shouldn't also support querying by FKey without the need for dot '.' notation. It seems some people expect the following syntax to work:

```js
  Pet.find({ "owner" : [1, 2] });
```

This does seem to work in some adapters apparently but it's not included in any tests or docs that I know of.

**Comment by stuk88** [comment]
+1

**Comment by atiertant** [comment]
we are thinking about adding this in deep populate (when merged) https://github.com/balderdashy/waterline/pull/1052
@particlebanana @devinivy filter results was our first idea but takes a lot of memory... this could be solved by doing something like this : 
with the previous exemple:

    Owner.findAll([{"pet.breed": "beagle"}, {billingMethod: "visa"}, {vet.state: "GA"}])

should become:

    vet.findAll({select:['Owner_id'],state: "GA"})         // => [55,56,57,58,80]
    pet.findAll({select:['Owner_id'],breed: "beagle"})   // => [25,32,57,58,205]

    Owner.findAll(id : [57,58],billingMethod: "visa"})

what do you think about this solution ? have you got a better one?

**Comment by devinivy** [comment]
The way I see this, Node.js is pretty good at I/O, and not so good at CPU-intensive tasks.  So the tradeoff of making more queries is fine with me.  I think there still may be a sweet spot between extra queries and extra filtering that we're not hitting yet.  This is a hard problem!  It would be great if there were some indicator which side of a many-to-many relationship is most efficient.  Perhaps we should take `dominant` sides into account here.

**Comment by atiertant** [comment]
@devinivy could you explain me what you would like to optimize in many-to-many ?
@devinivy if you wouldn't like CPU operations,and if @dmarcelino 's request is working,we could do request in series like this:

    Owner.findAll([{"pet.breed": "beagle"}, {billingMethod: "visa"}, {vet.state: "GA"}])

become

    vet.findAll({select:['Owner'],state: "GA"})         // => [55,56,57,58,80]
    pet.findAll({select:['Owner'],breed: "beagle",Owner: [55,56,57,58,80]}) // => [57,58]
    Owner.findAll(id : [57,58],billingMethod: "visa"})

of course it use less node cpu but request must be done in series so it takes more time...

**Comment by devinivy** [comment]
I mean, it depends on the size of those result sets to be able to tell which option takes more time.  There's also this: if the result set is so large that it's considered CPU/memory intensive, then do we really want to send that many ids in an `IN` query (as in the second example)?  The only thing worse than making node unhappy may be making your database unhappy.  This would be a good argument for the first option.

One way I could think of making your first option a little more performant would be to ask for the OwnerIds sorted by the database.  That would allow us to use a more efficient algorithm to perform the intersection of ids.

**Comment by atiertant** [comment]
your right https://lodash.com/docs#indexOf using a sorted array is faster.

**Comment by mhipo1364** [comment]
Hi Guys

I have a situation, I have `Content` model which has an `Owner`, and each `Owner` has many `Voucher`s

So I want to run query on content, which its owner has some particular voucher, How can I do that?

**Comment by devinivy** [comment]
You would currently do this in two queries: the first to find the owner of the voucher (`Voucher.find(id)`), and the second to find all `Content` with that owner (`Content.find({owner: voucherOwnerId})`).  In the future you may be able to use through associations to associate `Content` and `Vouchers` through `Owner`.  Waterline has room to grow for supporting these cross-table queries, but it is a non-trivial problem to solve.

**Comment by mhipo1364** [comment]
@devinivy :+1:

**Comment by atiertant** [comment]
@dmarcelino @devinivy is there any test on adapter for very large `IN` criteria ? if we use a very large criteria that result in a million of ids,and pass it as a `IN` criteria of a second request,must adapter give a result ? is there a limit?

**Comment by devinivy** [comment]
I don't know of any limit, nor any test!  Any limitations you run into should be from nodejs (i.e. max array size), the adapter you're using, or the database itself.  If a database has such a limit, ideally the adapter would ask for the results in pages.

**Comment by atiertant** [comment]
The exact maximum limit of an array is 2^32 - 1 or 4294967295, due to restrictions in Javascript's memory.
i think it's ok for what we wants to do but not sure adapters can handle it in `IN` criteria...
of course database has limit :  for exemple oracle can only handle 1000 value in `IN` clause but it could be solved by using a temporary table.
as waterline handle cross adapter operation, big request need large `IN` criteria on primary key for join.
there should be a waterline adapter spec.

**Comment by tjwebb** [comment]
It's up to each adapter to handle any idiosyncratic constraints, such as an `IN` array limit. Waterline needn't concern itself with such things.

**Comment by particlebanana** [comment]
You will more than likely run out of memory with the returned results before you hit any array limitations in javascript.

**Comment by atiertant** [comment]
@tjwebb yes but that's waterline who give spec.and what about errors?
@particlebanana handling 2^32 - 1 primaryKey in `IN` is silly(it was just to say that javascript was not the problem) but we need a minimal to work correctly with cross adapter operation : imagine you are searching the owners who have a dog called 'max' and having more than 30 years old , named 'john'.
they are maybe only about 50 owners but many many many dog called 'max' ...
if spec said adapters must handle 500 values in `IN` and there are 10000 dog called 'max' so we need 20 requests to solve it...

**Comment by sailsbot** [comment]
Thanks for posting, @gregorypierce.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
This is in a milestone, shouldn't have been closed!

**Comment by sailsbot** [comment]
Thanks for posting, @gregorypierce.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sailsbot** [comment]
Thanks for posting, @gregorypierce.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by particlebanana** [comment]
> Reopening until someone submits a PR to the roadmap for this feature.

**Comment by OwaysisApp** [comment]
Can someone make this happen? Sails has been so frustrating on this.  Been looking for a while and haven't found many resources that clarify the best way to do this.  :'/ Definitely taking the wind out of my sails.

**Comment by devinivy** [comment]
@OwaysisApp this is always possible using multiple queries.  It's a hard problem to solve across multiple connections/adapters, and waterline welcomes efforts towards achieving this!

**Comment by sailsbot** [comment]
Thanks for posting, @gregorypierce.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
This is indeed in roadmap.md, for those wondering :)

**Comment by mikermcneil** [comment]
@devinivy :+1:

**Comment by Dalanie** [comment]
It's one of the most important things developer need and the tickets exist since 13/14...

**Comment by andrejkosyak** [comment]
+

**Comment by 25matt12** [comment]
+1

**Comment by Oldwo1f** [comment]
I agree, this feature is really important

**Comment by marrouchi** [comment]
+1

**Comment by emps** [comment]
yeh i need this too

**Comment by daniyalme** [comment]
+1

**Comment by Simone-cogno** [comment]
+1

**Comment by Reda-bx** [comment]
+1

**Comment by pawelhertman** [comment]
+1

**Comment by ezemoro** [comment]
+1

**Comment by diegognt** [comment]
+1

**Comment by explicitcall** [comment]
+1

**Comment by ComicScrip** [comment]
+1

**Comment by alexmasita** [comment]
+1

**Comment by luislobo** [comment]
Current version supports it:

```javascript
User.find().populate('groups',{select:['name']}).exec(console.log)
```

**Comment by Crusader4Christ** [comment]
@luislobo Do you mean 1.0 or 0.12? In 0.12 it is not working

**Comment by luislobo** [comment]
I have just confirmed that it DOES NOT work with sails-mysql, but it DOES work with sails-mongo (that is what I was using on my tests).

I just created a project, using sails 0.12.11, waterline 0.11.6, using sails-mysql 0.11.5, and it didn't work:

```
sails> Group.find().populate('users').exec(console.log)
undefined
sails> null [ { users: 
     [ { name: 'Luis',
         id: 1,
         createdAt: '2017-01-12T16:33:41.000Z',
         updatedAt: '2017-01-12T16:33:41.000Z' },
       { name: 'Pablo',
         id: 2,
         createdAt: '2017-01-12T16:37:43.000Z',
         updatedAt: '2017-01-12T16:37:43.000Z' } ],
    name: 'Administrators',
    id: 1,
    createdAt: '2017-01-12T16:34:43.000Z',
    updatedAt: '2017-01-12T16:34:43.000Z' },
  { users: [],
    name: 'Normal Users',
    id: 2,
    createdAt: '2017-01-12T16:35:21.000Z',
    updatedAt: '2017-01-12T16:35:21.000Z' } ]

sails> Group.find().populate('users', {name:'Pablo'}).exec(console.log);
undefined
sails> null [ { users: 
     [ { name: 'Pablo',
         id: 2,
         createdAt: '2017-01-12T16:37:43.000Z',
         updatedAt: '2017-01-12T16:37:43.000Z' } ],
    name: 'Administrators',
    id: 1,
    createdAt: '2017-01-12T16:34:43.000Z',
    updatedAt: '2017-01-12T16:34:43.000Z' },
  { users: [],
    name: 'Normal Users',
    id: 2,
    createdAt: '2017-01-12T16:35:21.000Z',
    updatedAt: '2017-01-12T16:35:21.000Z' } ]

sails> Group.find().populate('users', {name:'Pablo', select:['name']}).exec(console.log)
undefined
sails> null [ { users: 
     [ { name: 'Pablo',
         id: 2,
         createdAt: '2017-01-12T16:37:43.000Z',
         updatedAt: '2017-01-12T16:37:43.000Z' } ],
    name: 'Administrators',
    id: 1,
    createdAt: '2017-01-12T16:34:43.000Z',
    updatedAt: '2017-01-12T16:34:43.000Z' },
  { users: [],
    name: 'Normal Users',
    id: 2,
    createdAt: '2017-01-12T16:35:21.000Z',
    updatedAt: '2017-01-12T16:35:21.000Z' } ]



```

But with sails-mongo, it did work:

```
sails> User.create({name:'Luis'}).exec(console.log)
undefined
sails> null { name: 'Luis',
  createdAt: '2017-01-12T16:43:23.774Z',
  updatedAt: '2017-01-12T16:43:23.774Z',
  id: '5877b22bcea9bdff625f6190' }
sails> User.create({name:'Pablo'}).exec(console.log)
undefined
sails> null { name: 'Pablo',
  createdAt: '2017-01-12T16:43:28.548Z',
  updatedAt: '2017-01-12T16:43:28.548Z',
  id: '5877b230cea9bdff625f6191' }
sails> Group.create({name:'Administrators'}).exec(console.log);
undefined
sails> null { name: 'Administrators',
  createdAt: '2017-01-12T16:43:35.694Z',
  updatedAt: '2017-01-12T16:43:35.694Z',
  id: '5877b237cea9bdff625f6192' }
sails> Group.create({name:'Normal Users', users:['5877b22bcea9bdff625f6190','5877b230cea9bdff625f6191']}).exec(console.log);
undefined
sails> null { name: 'Normal Users',
  createdAt: '2017-01-12T16:44:04.605Z',
  updatedAt: '2017-01-12T16:44:04.613Z',
  id: '5877b254cea9bdff625f6193' }
sails> Group.find().populate('users', {name:'Pablo', select:['name']}).exec(console.log)
undefined
sails> null [ { users: [],
    name: 'Administrators',
    createdAt: '2017-01-12T16:43:35.694Z',
    updatedAt: '2017-01-12T16:43:35.694Z',
    id: '5877b237cea9bdff625f6192' },
  { users: [ { name: 'Pablo', id: '5877b230cea9bdff625f6191' } ],
    name: 'Normal Users',
    createdAt: '2017-01-12T16:44:04.605Z',
    updatedAt: '2017-01-12T16:44:04.613Z',
    id: '5877b254cea9bdff625f6193' } ]

```

So, I would say it's a confirmed bug.

**Comment by particlebanana** [comment]
@luislobo `select` isn't officially supported in Sails `0.12`. The reason it works in Mongo is just due to the fact that it doesn't specifically strip it out.

In Sails `1.0` both `select` and `omit` are supported on all queries - top level and populated criteria.

**Comment by luislobo** [comment]
@particlebanana understood! Well, then we are lucky, MongoDB users. Looking forward to 1.0 then!

**Comment by Crusader4Christ** [comment]
@particlebanana Please, how can I add raw sql in where? Something like Group.find().populate('users').whereRawSql("user.name='Pablo') it is easy to implement and can be usefull in future.

**Comment by luislobo** [comment]
@Crusader4Christ you can use .query: http://sailsjs.com/documentation/reference/waterline-orm/models/query

**Comment by Crusader4Christ** [comment]
@luislobo Yes, I can use .query, but in this case I need to parse whole request by hands, eg:
```...
if (req.query.limit)...
if (req.query.skip)...
if (req.where)...
```
BUT what is the reason to use sails at all in this case? And what is the reason to parse this params if this can be done by sails? All I need is just adding parts of query in raw sql. This ability exists in most ORMs (knex.js, sequilize.js, Doctrine and Zend in php etc.)

As I know, your example with filtering in populate is working only not for MySQL.

**Comment by Crusader4Christ** [comment]
@luislobo filtering DOES NOT work with populate
`Group.find().populate('users', {name:'Pablo'}).exec(console.log);`
Pablo, Luis (mysql)

**Comment by jelhan** [comment]
@Crusader4Christ It does work in a certain way. Populate criteria filters which records are populated not which records are matched by find. So `Group.find().populate('users', { name: 'Pablo' })` will return all groups but only populate users named 'Pablo'.

This issue is about filtering records by attributes of associated records. This does not work yet. Since it's not mentioned in waterline oder sails changelog yet I won't expect it land in 1.0.

**Comment by Crusader4Christ** [comment]
@jelhan Any way to get raw sql from query object? Something like `sql = Model.find().where().toString()`?

**Comment by atiertant** [comment]
@Crusader4Christ do you want sql or a string like "Model.find().where()" ?

**Comment by davemackintosh** [comment]
I had to implement this in Multicolour via query strings, what I ended up having to do is create separate queries and then updating the query being made. The populate answers above didn't work for me in SQL based space so this is the best I could come up with. Here's my code for anyone interested in my solution.

https://github.com/Multicolour/multicolour/commit/61c8e438d37470a8b8e67f8fb45b12186e9a77d3 :+1:

**Comment by Mikeysax** [comment]
Wish this was implemented.

**Comment by simonmilz** [comment]
+1

**Comment by rafael-paiva** [comment]
+1

**Comment by ric2z** [comment]
+1

**Comment by glemiere** [comment]
+999999999

**Comment by bzupnick** [comment]
+1

**Comment by jarodsmk** [comment]
So, it's a couple years since this was opened - any news? Any feedback?

**Comment by rfrancois** [comment]
I'm waiting for this update too.

**Comment by HailoMYM** [comment]
+1

**Comment by glemiere** [comment]
I made a small class to connect a ionic app to a sailsjs app (so it does use waterline).
One function in there could be re-written server side to finally fix this issue.

I don't really have the time right now but if anybody wants to take a look at the gist below, the `remove` function takes an association object parameter with two keys : `where` (the id or query) and `what` (the model to query) and does the job of building a request to remove relationships. The same concept can be adapted for GET requests wether it's on client or server side.

Hope it helps!
 
https://gist.github.com/glemiere/426e2ada35a775d22a92819122798c5e#file-sailsapicall-ts-L226

**Comment by geeckmc** [comment]
+1

**Comment by grantbi** [comment]
Is the best current approach to this to query for all (limit:Number.MAX_SAFE_INTEGER) with a `populate('model', {where:_____}` and then filter the results?

Or did I misunderstand something?

**Comment by agustindaguerre** [comment]
+1
